phoenix-core/src/test/java/org/apache/phoenix/query/PhoenixStatsCacheLoaderTest.java

```diff
@@ -0,0 +1,123 @@
+     */
+    protected class TestStatsLoaderImpl implements PhoenixStatsLoader {
+        private int maxLength = 1;
+
+        @Override
+        public boolean needsLoad() {
+            // Whenever it's called, we try to load stats from stats table
+            // no matter it has been updated or not.
+            return true;
+        }
+
+        @Override
+        public GuidePostsInfo loadStats(GuidePostsKey statsKey, GuidePostsInfo prevGuidepostInfo) throws Exception {
+            try {
+                Thread.sleep(2000);
+            }
+            catch (InterruptedException e) {
+                assertFalse(true);
+            }
+
+            return new GuidePostsInfo(Collections.<Long> emptyList(),
+                    new ImmutableBytesWritable(ByteUtil.EMPTY_BYTE_ARRAY),
+                    Collections.<Long> emptyList(), maxLength++, 0, Collections.<Long> emptyList());
+        }
+    }
+
+    GuidePostsInfo getStats(LoadingCache<GuidePostsKey, GuidePostsInfo> cache, GuidePostsKey guidePostsKey) {
+        GuidePostsInfo guidePostsInfo;
+        try {
+            guidePostsInfo = cache.get(guidePostsKey);
+        } catch (ExecutionException e) {
+            assertFalse(true);
+            return GuidePostsInfo.NO_GUIDEPOST;
+        }
+
+        return guidePostsInfo;
+    }
+
+    void sleep(int x) {
+        try {
+            Thread.sleep(x);
+        }
+        catch (InterruptedException e) {
+            assertFalse(true);
+        }
+    }
+
+    @Test
+    public void testStatsBeingAutomaticallyRefreshed() {
+        ExecutorService executor = Executors.newFixedThreadPool(4);
+
+        LoadingCache<GuidePostsKey, GuidePostsInfo> cache = CacheBuilder.newBuilder()
+                // Refresh entries a given amount of time after they were written
+                .refreshAfterWrite(1000, TimeUnit.MILLISECONDS)
+                // Maximum total weight (size in bytes) of stats entries
+                .maximumWeight(QueryServicesOptions.DEFAULT_STATS_MAX_CACHE_SIZE)
+                // Defer actual size to the PTableStats.getEstimatedSize()
+                .weigher(new Weigher<GuidePostsKey, GuidePostsInfo>() {
+                    @Override public int weigh(GuidePostsKey key, GuidePostsInfo info) {
+                        return info.getEstimatedSize();
+                    }
+                })
+                // Log removals at TRACE for debugging
+                .removalListener(new GuidePostsCache.PhoenixStatsCacheRemovalListener())
+                // Automatically load the cache when entries are missing
+                .build(new PhoenixStatsCacheLoader(new TestStatsLoaderImpl(), executor));
+
+        GuidePostsKey guidePostsKey = new GuidePostsKey(new byte[4], new byte[4]);
+        GuidePostsInfo guidePostsInfo = getStats(cache, guidePostsKey);
+        assertTrue(guidePostsInfo.getMaxLength() == 1);
+        sleep(3000);
+        guidePostsInfo = getStats(cache, guidePostsKey);
```

Comment:
Comment from @twdsilva (copied from prev PR #416): Instead of relying on sleep which tends to cause tests to fail on the build machines, can you use a latch or some other mechanism to determine that the guideposts have been refreshed?

Reply:
You can use this instead of actual time to make tests faster and reliable. 
https://stackoverflow.com/questions/29990788/guava-ticker-cache-expire

Reply:
Fixed by using CountDownLatch to guarantee deterministic test behavior.